From 13e8a13539fdf12065c4754261f08b0d23536652 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Mon, 6 May 2024 17:42:34 +0800
Subject: [PATCH 113/137] mtk: hostapd: support capi ap_rfeatures trig_variant
 for wifi7 cert

Support new hostapd_cli command as below:
$ hostapd_cli -i <intf> raw ap_rfeatures trig_variant=<type>

This will prepare nl80211 msg
MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_VARIANT_TYPE and then send to driver
via nl80211.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 hostapd/ctrl_iface.c         | 2 ++
 src/common/mtk_vendor.h      | 1 +
 src/drivers/driver_nl80211.c | 1 +
 3 files changed, 4 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index d7d1a0159..3708b67a0 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -4930,6 +4930,8 @@ hostapd_ctrl_iface_ap_rfeatures(struct hostapd_data *hapd, char *cmd,
 		goto trigtype;
 	} else if (os_strcmp(config, "ack_policy") == 0)
 		sub_cmd = MTK_VENDOR_ATTR_RFEATURE_CTRL_ACK_PLCY;
+	else if (os_strcmp(config, "trig_variant") == 0)
+		sub_cmd = MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_VARIANT_TYPE;
 	else {
 		wpa_printf(MSG_ERROR,
 			"Unsupported parameter %s for ap_rfeatures", config);
diff --git a/src/common/mtk_vendor.h b/src/common/mtk_vendor.h
index 777b4c395..71c1e5b83 100644
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -191,6 +191,7 @@ enum mtk_vendor_attr_rfeature_ctrl {
 	MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_TYPE,
 	MTK_VENDOR_ATTR_RFEATURE_CTRL_ACK_PLCY,
 	MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_TXBF,
+	MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_VARIANT_TYPE,
 
 	/* keep last */
 	NUM_MTK_VENDOR_ATTRS_RFEATURE_CTRL,
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 5fee13236..6f251f6d5 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -126,6 +126,7 @@ rfeature_ctrl_policy[NUM_MTK_VENDOR_ATTRS_RFEATURE_CTRL] = {
 	[MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_TYPE] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_RFEATURE_CTRL_ACK_PLCY] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_TXBF] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_RFEATURE_CTRL_TRIG_VARIANT_TYPE] = { .type = NLA_U8 },
 };
 
 static const struct nla_policy
-- 
2.18.0

